components/aws/sagemaker/codebuild/integration-test.buildspec.yml

```diff
@@ -0,0 +1,18 @@
+version: 0.2      
+phases:
+  build:
+    commands:
+      - cd components/aws
+      - docker build . -f ./sagemaker/tests/integration_tests/Dockerfile -t amazon/integration-test-image --quiet
+
+      # Run the container and copy the results to /tmp
+      # Passes all host environment variables through to the container
+      - docker run --name integration-test-container $(env | cut -f1 -d= | sed 's/^/-e /') amazon/integration-test-image
```

Comment:
Design question: 

1. What is the need to run container in container here ?

- We can have a codebuild step which builds all the images needed for the pipeline(release/unit/integ-test) first so that each step can use those as base image of codebuild job directly
 Especially with integration test and deploy - the test container needs IAM permissions, and maybe things where it need access to resources outside the codebuild job so we can use avoid passing them explicitly here.
Also we can maintain only one image per job rather than a base image + step specific image 

2. Placement of these codebuild spec files
 - Why do we need a separate codebuild directory, can we add the specs into their respective directories? create a release directory if needed

Reply:
>     * We can have a codebuild step which builds all the images needed for the pipeline(release/unit/integ-test) first so that each step can use those as base image of codebuild job directly
>       Especially with integration test and deploy - the test container needs IAM permissions, and maybe things where it need access to resources outside the codebuild job so we can use avoid passing them explicitly here.
>       Also we can maintain only one image per job rather than a base image + step specific image

So that was the way we designed the operators, but it turns out to be difficult from a developer experience. By doing that, we needed to create a local version of CodeBuild to emulate the environment, and then inside that we injected the integration test scripts. That ends up being unnecessary overhead and adds complexity that doesn't need to be there. Plus now we have to maintain a second image which becomes confusing about where certain resources are defined.

My ideal world would be this:
- The developer of a new integration test can edit a `.env` file and then build and run the integration test Dockerfile. Everything would be created automatically (except data maybe) and all of the tools needed for testing would be included in that Docker container.
- All the pipeline needs to do is build and run the same container. The scripts would be written in such a way that we could programatically override any values (role names, bucket names, etc.) to work with either system.

I have already included in this CodeBuild spec the ability for the outer CodeBuild container to pass all of its environment variables through to the inner (integration test) container. This allows us to simply pass any values in from our CDK infra to the CodeBuild container and the integration test container would be able to access them - this includes role permissions.

Reply:
>     * Why do we need a separate codebuild directory, can we add the specs into their respective directories? create a release directory if needed

It is fairly common for projects to have a "codebuild" directory, but I've seen both solutions. I think because each of these codebuild steps does roughly the same thing (build and run a container), they should stay fairly close together so we remember to update them. They also don't really have anything to do with the actual unit or integration tests themselves - I wouldn't expect anyone else to care about them except us.